src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs

```diff
@@ -1019,446 +794,420 @@ public bool Overlaps(IEnumerable<T> other)
-                        throw new InvalidOperationException(SR.InvalidOperation_ConcurrentOperationsNotSupported);
+                        ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
                     }
-                    collisionCount++;
                 }
             }
 
             int index;
-            if (_freeList >= 0)
+            if (_freeCount > 0)
             {
                 index = _freeList;
-                _freeList = slots[index].next;
+                _freeCount--;
+                Debug.Assert((StartOfFreeList - entries![_freeList].Next) >= -1, "shouldn't overflow because `next` cannot underflow");
+                _freeList = StartOfFreeList - entries[_freeList].Next;
             }
             else
             {
-                if (_lastIndex == slots.Length)
+                int count = _count;
+                if (count == entries.Length)
                 {
-                    IncreaseCapacity();
-                    // this will change during resize
-                    slots = _slots;
-                    bucket = hashCode % _buckets.Length;
+                    Resize();
+                    bucket = ref GetBucketRef(hashCode);
                 }
-                index = _lastIndex;
-                _lastIndex++;
-            }
-            slots[index].hashCode = hashCode;
-            slots[index].value = value;
-            slots[index].next = _buckets[bucket] - 1;
-            _buckets[bucket] = index + 1;
-            _count++;
-            _version++;
-
-            return true;
-        }
+                index = count;
+                _count = count + 1;
+                entries = _entries;
+            }
 
-        // Add value at known index with known hash code. Used only
-        // when constructing from another HashSet.
-        private void AddValue(int index, int hashCode, T value)
-        {
-            int bucket = hashCode % _buckets!.Length;
+            {
+                ref Entry entry = ref entries![index];
+                entry.HashCode = hashCode;
+                entry.Next = bucket - 1; // Value in _buckets is 1-based
+                entry.Value = value;
+                bucket = index + 1;
+                _version++;
+                location = index;
+            }
 
-#if DEBUG
-            IEqualityComparer<T> comparer = _comparer ?? EqualityComparer<T>.Default;
-            Debug.Assert(InternalGetHashCode(value, comparer) == hashCode);
-            for (int i = _buckets[bucket] - 1; i >= 0; i = _slots[i].next)
+            // Value types never rehash
+            if (!typeof(T).IsValueType && collisionCount > HashHelpers.HashCollisionThreshold && comparer is NonRandomizedStringEqualityComparer)
             {
-                Debug.Assert(!comparer.Equals(_slots[i].value, value));
+                // If we hit the collision threshold we'll need to switch to the comparer which is using randomized string hashing
```

Comment:
BTW, I'm not sure whether we test that binary serialization (boo) works for both secure and non secure string comparers. I do see a dictionary test with a comment asserting that it should be invisible:
```c#
        [Fact]
        public void ComparerSerialization()
        {
            // Strings switch between randomized and non-randomized comparers,
            // however this should never be observable externally.
            TestComparerSerialization(EqualityComparer<string>.Default);
```
Anyway - if we have such a test, it should possibly exist for both.